### PR TITLE
batt_smbus: A few fixes with BQ40Z80

### PR DESCRIPTION
**Pull request by amikhalev**
**Describe problem solved by this pull request**
`batt_smbus` running with BQ40Z80 would produce an error from [this line](https://github.com/PX4/PX4-Autopilot/blob/e48a86916006794989cda3d094cfb7a31258f21d/platforms/common/i2c_spi_buses.cpp#L510).

It also reports errors in `perf` under `drivers__smbus: errors:` and does not publish `battery_status`.

**Describe your solution**
Pass device ID to `I2CSPIDriver`.

After some inspection I determined the second problem was due to the PEC failing because `SMBus::read_block` was called with a length of 34. The actual data is only 32 bytes, but devices sends back the 16-bit address of the block before sending the data, so it ends up being 34 bytes. So increase the max buffer size to 34.

I couldn't find a maximum size for this after a quick glance of the SMBus spec, so I suppose for more generality it could be increased further, but for now it's probably better to conserve stack usage.

**Test data / coverage**
Tested on Holybro Durandal with BQ40Z80. No smbus errors, and does publish `battery_status`.

### CODE REVIEW

**Comment by mhkabir** [comment]
@amikhalev can you also add my uORB advertisement context fix here? Otherwise this driver only works if the battery happens to be connected on the last external bus in the probe list.

**Comment by mhkabir** [comment]
@dagar there are more things weird with this driver like hackily getting and _setting_ some of the deprecated params (e.g BAT_SOURCE, BAT_N_CELLS). 

Instead of patching this, we're currently working on a dedicated, cleaned up bq40z80 driver which will make its way upstream shortly.

**Comment by dagar** [comment]
> Instead of patching this, we're currently working on a dedicated, cleaned up bq40z80 driver which will make its way upstream shortly.

The other thing that's been badly overloaded is `battery_status` and should probably be split for smart batteries. The configuration and handling of multiple instances of smart batteries is something I'd like to get under control.

**Comment by echoGee** [comment]
@amikhalev @mhkabir Could you build the z80 driver based on some of the base classes from this PR: https://github.com/PX4/PX4-Autopilot/pull/16723 ?

**Comment by mhkabir** [comment]
@echoGee happy to; can you guys split out the base smbus implementation out to a separate PR?

**Comment by echoGee** [comment]
> @echoGee happy to; can you guys split out the base smbus implementation out to a separate PR?

We can do that. But @dagar suggests that the different devices running _SMBUS-based_ (batmon, bq40z80 etc) be different modules. His suggestion was to prevent regression on new PX4 releases. We are trying to go down the path of making batmon a separate module and end up with too much duplicated code. I worry that there would be fixes that make it on either branches that do not make it to the other. 

I'm waiting on some feedback for this request: 
https://github.com/PX4/PX4-Autopilot/pull/16723#issuecomment-780774213
Please chime in if you have any opinions @mhkabir

**Comment by BazookaJoe1900** [comment]
@amikhalev about read_block() I noticed the issue too. On my repo its also been changed to 34 bytes, and I see no issues since.
I had no time to push that, sorry...

about the general structure of the driver. I agree that there many enhancements, fixes and restructures that can be done.
from my side, I can say that
- sealing/unsealing should be only optional (maybe by parameter
- the driver should not change BAT_SOURCE. it either smart battery should be assemble, or not. if there is no comm, there should be error to the operator
- Auto detection Z50 vs Z80.

**Review by dagar — APPROVED** [review]

File: src/drivers/batt_smbus/batt_smbus.cpp [hunk 1]
```diff
@@ -192,7 +190,9 @@ void BATT_SMBUS::RunImpl()
 		}
 
 		new_report.interface_error = perf_event_count(_interface->_interface_errors);
-		orb_publish(ORB_ID(battery_status), _batt_topic, &new_report);
+
+		int instance = 0;
+		orb_publish_auto(ORB_ID(battery_status), &_batt_topic, &new_report, &instance);
```
Comment (dagar) [inline src/drivers/batt_smbus/batt_smbus.cpp:195]: We can make _batt_topic a uORB::PublicationMulti<battery_status_s> later.

**Pull request merged**